File: go/analyze/rules_types.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package analyze

import (
	"context"
	"regexp"
	"strings"

	keystonev1alpha1 "github.com/dogukanturhal/keystone/api/v1alpha1"
)

// Phase A2 — PostgreSQL type and schema-declaration conventions.
//
// These rules push authors toward the modern PG types that every
// production team eventually converges on. Calibrated as warnings /
// notices — they are not blockers, but persistent violation signals an
// author unaware of the PG edge cases.

// -- Rule: no-varchar-without-limit ------------------------------------

// NoVarcharWithoutLimit flags VARCHAR (a.k.a. CHARACTER VARYING) used
// without a length specifier. Unbounded VARCHAR has the same storage as
// TEXT but carries a length-check bit of overhead on assignment. The
// PostgreSQL documentation explicitly recommends TEXT when no bound is
// needed and VARCHAR(N) when there is one.
type NoVarcharWithoutLimit struct{}

func (NoVarcharWithoutLimit) ID() string { return "no-varchar-without-limit" }
func (NoVarcharWithoutLimit) Description() string {
	return "unbounded VARCHAR has no advantage over TEXT; be explicit"
}

// RE2 has no negative lookahead; use the two-regex pattern already
// established for timestamp-vs-timestamptz. reVarcharAny matches every
// occurrence; reVarcharBounded matches the bounded `VARCHAR(N)` form —
// we flag lines where the former matches and the latter does not.
var reVarcharAny = regexp.MustCompile(`(?i)\b(VARCHAR|CHARACTER\s+VARYING)\b`)
var reVarcharBounded = regexp.MustCompile(`(?i)\b(VARCHAR|CHARACTER\s+VARYING)\s*\(`)

func (a *NoVarcharWithoutLimit) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		for i, line := range strings.Split(f.Body, "\n") {
			if !reVarcharAny.MatchString(line) {
				continue
			}
			// If every VARCHAR on this line is followed by `(`, the
			// count of bounded matches equals the count of total
			// matches — nothing to flag. We re-scan with the bounded
			// regex because a single line may mix both forms.
			anyCount := len(reVarcharAny.FindAllStringIndex(line, -1))
			boundedCount := len(reVarcharBounded.FindAllStringIndex(line, -1))
			if boundedCount >= anyCount {
				continue
			}
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelNotice,
				File:     f.Name,
				Line:     int32(i + 1),
				Message: "unbounded VARCHAR / CHARACTER VARYING has no advantage over TEXT. " +
					"Use TEXT when the length is unbounded, VARCHAR(N) when it isn't.",
			})
		}
	}
	return out, nil
}

// -- Rule: prefer-jsonb-over-json --------------------------------------

// PreferJsonbOverJson flags `json` column type in favour of `jsonb`.
// `json` stores the original text verbatim — no indexing, no
// normalization, repeated key lookups re-parse on every query.
// `jsonb` stores a decomposed binary form, supports GIN indexes, and
// is the right default for almost every new column.
type PreferJsonbOverJson struct{}

func (PreferJsonbOverJson) ID() string { return "prefer-jsonb-over-json" }
func (PreferJsonbOverJson) Description() string {
	return "use JSONB for indexability and parse-once storage; JSON only for strict round-trip"
}

// Match `json` as a column type only — `json` can appear elsewhere
// (comments, function names, identifier "json"). We anchor on the
// "<name> json" pattern inside DDL: column name followed by
// whitespace followed by the word JSON. `\bJSON\b` does NOT match the
// leading JSON of JSONB (J→S→O→N→B is all word chars, so there's no
// word boundary between N and B) — no negative-lookahead needed.
var reJsonType = regexp.MustCompile(`(?i)\b\w+\s+JSON\b`)

// Narrow the match to DDL contexts only — don't flag SELECT using
// json_build_object etc. A quick and cheap check: skip files that
// have no CREATE TABLE / ADD COLUMN / ALTER COLUMN.
func hasColumnDeclaration(body string) bool {
	return regexp.MustCompile(`(?i)\b(CREATE\s+TABLE|ADD\s+COLUMN|ALTER\s+COLUMN)\b`).MatchString(body)
}

func (a *PreferJsonbOverJson) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		if !hasColumnDeclaration(f.Body) {
			continue
		}
		for i, line := range strings.Split(f.Body, "\n") {
			if !reJsonType.MatchString(line) {
				continue
			}
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelNotice,
				File:     f.Name,
				Line:     int32(i + 1),
				Message: "column typed as JSON. JSONB is binary-decomposed, indexable " +
					"(GIN / path-ops), and parses once at INSERT. JSON preserves " +
					"whitespace + key order — useful only for strict byte-round-trip.",
			})
		}
	}
	return out, nil
}

// -- Rule: no-numeric-without-precision --------------------------------

// NoNumericWithoutPrecision flags NUMERIC / DECIMAL columns declared
// without a precision spec. Unbounded NUMERIC has variable-precision
// arbitrary-digit storage; it's slow, big, and almost always wrong
// for a column that represents "money", "percentage", or any real-
// world metric. Explicit NUMERIC(p, s) forces the author to think
// about the range.
type NoNumericWithoutPrecision struct{}

func (NoNumericWithoutPrecision) ID() string { return "no-numeric-without-precision" }
func (NoNumericWithoutPrecision) Description() string {
	return "NUMERIC without precision is arbitrary-digit; declare (p, s) explicitly"
}

// Same two-regex pattern as VARCHAR.
var reNumericAny = regexp.MustCompile(`(?i)\b\w+\s+(NUMERIC|DECIMAL)\b`)
var reNumericBounded = regexp.MustCompile(`(?i)\b\w+\s+(NUMERIC|DECIMAL)\s*\(`)

func (a *NoNumericWithoutPrecision) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		if !hasColumnDeclaration(f.Body) {
			continue
		}
		for i, line := range strings.Split(f.Body, "\n") {
			if !reNumericAny.MatchString(line) {
				continue
			}
			anyCount := len(reNumericAny.FindAllStringIndex(line, -1))
			boundedCount := len(reNumericBounded.FindAllStringIndex(line, -1))
			if boundedCount >= anyCount {
				continue
			}
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelNotice,
				File:     f.Name,
				Line:     int32(i + 1),
				Message: "unbounded NUMERIC / DECIMAL stores arbitrary-precision digits — slow, " +
					"heavy, and usually wrong. Declare the intended range: " +
					"NUMERIC(12, 2) for money, NUMERIC(5, 4) for probabilities, etc.",
			})
		}
	}
	return out, nil
}

// -- Rule: require-if-not-exists-on-create-table -----------------------

// RequireIfNotExistsOnCreateTable flags CREATE TABLE without IF NOT
// EXISTS. Under GitOps + retry loops, a migration may be re-applied
// after a partial failure; IF NOT EXISTS makes CREATE TABLE idempotent
// so the retry doesn't error. Not universally required — some shops
// prefer strict mode — so this is a Notice.
type RequireIfNotExistsOnCreateTable struct{}

func (RequireIfNotExistsOnCreateTable) ID() string { return "require-if-not-exists-on-create-table" }
func (RequireIfNotExistsOnCreateTable) Description() string {
	return "CREATE TABLE IF NOT EXISTS makes migrations idempotent under retry"
}

var reCreateTableStrict = regexp.MustCompile(`(?i)\bCREATE\s+TABLE\s+(\w+)\b`)
var reCreateTableIfNotExists = regexp.MustCompile(`(?i)\bCREATE\s+TABLE\s+IF\s+NOT\s+EXISTS\b`)

func (a *RequireIfNotExistsOnCreateTable) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		for i, line := range strings.Split(f.Body, "\n") {
			if !reCreateTableStrict.MatchString(line) {
				continue
			}
			if reCreateTableIfNotExists.MatchString(line) {
				continue
			}
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelNotice,
				File:     f.Name,
				Line:     int32(i + 1),
				Message: "CREATE TABLE without IF NOT EXISTS. Retries after partial failure " +
					"re-run the whole file — add IF NOT EXISTS to make creation idempotent.",
			})
		}
	}
	return out, nil
}

// -- Rule: no-uuid-generate-v1 -----------------------------------------

// NoUuidGenerateV1 flags uuid_generate_v1() / v1mc(). V1 UUIDs encode
// the host MAC address and a timestamp — leaking both in every row's
// primary key. Since PG 13, `gen_random_uuid()` is in core and
// produces v4 UUIDs with no leak.
type NoUuidGenerateV1 struct{}

func (NoUuidGenerateV1) ID() string { return "no-uuid-generate-v1" }
func (NoUuidGenerateV1) Description() string {
	return "uuid_generate_v1 leaks host MAC + timestamp; use gen_random_uuid()"
}

var reUuidV1 = regexp.MustCompile(`(?i)\buuid_generate_v1(mc)?\s*\(`)

func (a *NoUuidGenerateV1) Check(_ context.Context, m *Migration) ([]Finding, error) {
	return statementMatches(m, reUuidV1, keystonev1alpha1.LintLevelWarning,
		a.ID(),
		"uuid_generate_v1 / v1mc encode host MAC address + timestamp, leaking both "+
			"in every row's identifier. On PG 13+, gen_random_uuid() produces "+
			"v4 UUIDs with no leakage and no extension required."), nil
}

File: rust/keystone-sdk/src/analyze/rules_types.rs
// SPDX-License-Identifier: Apache-2.0

//! Phase A2 — PG type and schema-declaration conventions. Ported from
//! `rules_types.go`.

use std::sync::LazyLock;

use regex::Regex;

use super::{finding, statement_matches, Analyzer, Finding, LintLevel, Migration};

macro_rules! re {
    ($name:ident, $pat:expr) => {
        static $name: LazyLock<Regex> = LazyLock::new(|| Regex::new($pat).expect($pat));
    };
}

re!(RE_VARCHAR_ANY, r"(?i)\b(VARCHAR|CHARACTER\s+VARYING)\b");
re!(
    RE_VARCHAR_BOUNDED,
    r"(?i)\b(VARCHAR|CHARACTER\s+VARYING)\s*\("
);
re!(RE_JSON_TYPE, r"(?i)\b\w+\s+JSON\b");
re!(
    RE_HAS_COLUMN_DECLARATION,
    r"(?i)\b(CREATE\s+TABLE|ADD\s+COLUMN|ALTER\s+COLUMN)\b"
);
re!(RE_NUMERIC_ANY, r"(?i)\b\w+\s+(NUMERIC|DECIMAL)\b");
re!(RE_NUMERIC_BOUNDED, r"(?i)\b\w+\s+(NUMERIC|DECIMAL)\s*\(");
re!(RE_CREATE_TABLE_STRICT, r"(?i)\bCREATE\s+TABLE\s+(\w+)\b");
re!(
    RE_CREATE_TABLE_IF_NOT_EXISTS,
    r"(?i)\bCREATE\s+TABLE\s+IF\s+NOT\s+EXISTS\b"
);
re!(RE_UUID_V1, r"(?i)\buuid_generate_v1(mc)?\s*\(");

fn has_column_declaration(body: &str) -> bool {
    RE_HAS_COLUMN_DECLARATION.is_match(body)
}

/// no-varchar-without-limit
pub struct NoVarcharWithoutLimit;
impl Analyzer for NoVarcharWithoutLimit {
    fn id(&self) -> &'static str {
        "no-varchar-without-limit"
    }
    fn description(&self) -> &'static str {
        "unbounded VARCHAR has no advantage over TEXT; be explicit"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            for (i, line) in f.body.split('\n').enumerate() {
                if !RE_VARCHAR_ANY.is_match(line) {
                    continue;
                }
                let any_count = RE_VARCHAR_ANY.find_iter(line).count();
                let bounded_count = RE_VARCHAR_BOUNDED.find_iter(line).count();
                if bounded_count >= any_count {
                    continue;
                }
                out.push(finding(
                    self.id(),
                    LintLevel::Notice,
                    &f.name,
                    (i + 1) as i32,
                    "unbounded VARCHAR / CHARACTER VARYING has no advantage over TEXT. \
                     Use TEXT when the length is unbounded, VARCHAR(N) when it isn't.",
                ));
            }
        }
        out
    }
}

/// prefer-jsonb-over-json
pub struct PreferJsonbOverJson;
impl Analyzer for PreferJsonbOverJson {
    fn id(&self) -> &'static str {
        "prefer-jsonb-over-json"
    }
    fn description(&self) -> &'static str {
        "use JSONB for indexability and parse-once storage; JSON only for strict round-trip"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            if !has_column_declaration(&f.body) {
                continue;
            }
            for (i, line) in f.body.split('\n').enumerate() {
                if !RE_JSON_TYPE.is_match(line) {
                    continue;
                }
                out.push(finding(
                    self.id(),
                    LintLevel::Notice,
                    &f.name,
                    (i + 1) as i32,
                    "column typed as JSON. JSONB is binary-decomposed, indexable \
                     (GIN / path-ops), and parses once at INSERT. JSON preserves \
                     whitespace + key order — useful only for strict byte-round-trip.",
                ));
            }
        }
        out
    }
}

/// no-numeric-without-precision
pub struct NoNumericWithoutPrecision;
impl Analyzer for NoNumericWithoutPrecision {
    fn id(&self) -> &'static str {
        "no-numeric-without-precision"
    }
    fn description(&self) -> &'static str {
        "NUMERIC without precision is arbitrary-digit; declare (p, s) explicitly"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            if !has_column_declaration(&f.body) {
                continue;
            }
            for (i, line) in f.body.split('\n').enumerate() {
                if !RE_NUMERIC_ANY.is_match(line) {
                    continue;
                }
                let any_count = RE_NUMERIC_ANY.find_iter(line).count();
                let bounded_count = RE_NUMERIC_BOUNDED.find_iter(line).count();
                if bounded_count >= any_count {
                    continue;
                }
                out.push(finding(
                    self.id(),
                    LintLevel::Notice,
                    &f.name,
                    (i + 1) as i32,
                    "unbounded NUMERIC / DECIMAL stores arbitrary-precision digits — slow, \
                     heavy, and usually wrong. Declare the intended range: \
                     NUMERIC(12, 2) for money, NUMERIC(5, 4) for probabilities, etc.",
                ));
            }
        }
        out
    }
}

/// require-if-not-exists-on-create-table
pub struct RequireIfNotExistsOnCreateTable;
impl Analyzer for RequireIfNotExistsOnCreateTable {
    fn id(&self) -> &'static str {
        "require-if-not-exists-on-create-table"
    }
    fn description(&self) -> &'static str {
        "CREATE TABLE IF NOT EXISTS makes migrations idempotent under retry"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            for (i, line) in f.body.split('\n').enumerate() {
                if !RE_CREATE_TABLE_STRICT.is_match(line) {
                    continue;
                }
                if RE_CREATE_TABLE_IF_NOT_EXISTS.is_match(line) {
                    continue;
                }
                out.push(finding(
                    self.id(),
                    LintLevel::Notice,
                    &f.name,
                    (i + 1) as i32,
                    "CREATE TABLE without IF NOT EXISTS. Retries after partial failure \
                     re-run the whole file — add IF NOT EXISTS to make creation idempotent.",
                ));
            }
        }
        out
    }
}

/// no-uuid-generate-v1
pub struct NoUuidGenerateV1;
impl Analyzer for NoUuidGenerateV1 {
    fn id(&self) -> &'static str {
        "no-uuid-generate-v1"
    }
    fn description(&self) -> &'static str {
        "uuid_generate_v1 leaks host MAC + timestamp; use gen_random_uuid()"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        statement_matches(
            m,
            &RE_UUID_V1,
            LintLevel::Warning,
            self.id(),
            "uuid_generate_v1 / v1mc encode host MAC address + timestamp, leaking both \
             in every row's identifier. On PG 13+, gen_random_uuid() produces \
             v4 UUIDs with no leakage and no extension required.",
        )
    }
}

#[cfg(test)]
mod tests {
    use super::super::{FileBody, Migration};
    use super::*;

    fn mig(body: &str) -> Migration {
        Migration {
            files: vec![FileBody {
                name: "001.sql".into(),
                body: body.into(),
            }],
            ..Default::default()
        }
    }

    #[test]
    fn varchar_limit() {
        assert_eq!(NoVarcharWithoutLimit.check(&mig("a VARCHAR")).len(), 1);
        assert_eq!(NoVarcharWithoutLimit.check(&mig("a VARCHAR(20)")).len(), 0);
    }

    #[test]
    fn jsonb_preference_gated_on_ddl() {
        assert_eq!(
            PreferJsonbOverJson
                .check(&mig("CREATE TABLE t (a json);"))
                .len(),
            1
        );
        // JSONB does not match (no word boundary N→B).
        assert_eq!(
            PreferJsonbOverJson
                .check(&mig("CREATE TABLE t (a jsonb);"))
                .len(),
            0
        );
        // Not a DDL context → skipped.
        assert_eq!(
            PreferJsonbOverJson
                .check(&mig("SELECT json_build_object('a', 1);"))
                .len(),
            0
        );
    }

    #[test]
    fn numeric_precision() {
        assert_eq!(
            NoNumericWithoutPrecision
                .check(&mig("CREATE TABLE t (a NUMERIC);"))
                .len(),
            1
        );
        assert_eq!(
            NoNumericWithoutPrecision
                .check(&mig("CREATE TABLE t (a NUMERIC(12,2));"))
                .len(),
            0
        );
    }

    #[test]
    fn create_table_if_not_exists() {
        assert_eq!(
            RequireIfNotExistsOnCreateTable
                .check(&mig("CREATE TABLE t (id int);"))
                .len(),
            1
        );
        assert_eq!(
            RequireIfNotExistsOnCreateTable
                .check(&mig("CREATE TABLE IF NOT EXISTS t (id int);"))
                .len(),
            0
        );
    }

    #[test]
    fn uuid_v1() {
        assert_eq!(
            NoUuidGenerateV1
                .check(&mig("DEFAULT uuid_generate_v1()"))
                .len(),
            1
        );
        assert_eq!(
            NoUuidGenerateV1
                .check(&mig("DEFAULT gen_random_uuid()"))
                .len(),
            0
        );
    }
}
